book/src/dev/rfcs/0003-state-updates.md

```diff
@@ -0,0 +1,305 @@
+that's required to process some request asynchronously is to clone the
+appropriate handle, move it into an async block, and make the call as part of
+the future. (We might want to use Tokio's blocking API for this, but this is
+an implementation detail).
+
+Because the state service has exclusive access to the sled database, and the
+state service sees a linearized sequence of state requests, we have an easy
+way to opt in to asynchronous database access. We can perform sled operations
+synchronously in the `Service::call`, waiting for them to complete, and be
+sure that all future requests will see the resulting sled state. Or, we can
+perform sled operations asynchronously in the future returned by
+`Service::call`.
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+In summary:
+
+- **Sled reads** may be done synchronously (in `call`) or asynchronously (in
+  the `Future`), depending on the context;
+
+- **Sled writes** must be done synchronously (in `call`).
+
+## In-memory data structures
+[in-memory]: #in-memory
+
+At a high level, the in-memory data structures store a collection of chains,
+each rooted at the highest finalized block. Each chain consists of a map from
+heights to blocks. Chains are stored using an ordered map from difficulty to
+chains, so that the map ordering is the ordering of best to worst chains.
+
+- XXX fill in details on exact types
+
+- XXX work out whether we should store extra data (e.g., a HashSet of UTXOs
+  spent by some block etc.) to speed up checks.
+
+When a new block extends the best chain past 100 blocks, the old root is
+removed from the in-memory state and committed to sled.
+
+## Sled data structures
+[sled]: #sled
+
+Sled provides a persistent, thread-safe `BTreeMap<&[u8], &[u8]>`. Each map is
+a distinct "tree". Keys are sorted using lex order on byte strings, so
+integer values should be stored using big-endian encoding (so that the lex
+order on byte strings is the numeric ordering).
+
+We use the following Sled trees:
+
+| Tree                 |                  Keys |                              Values |
+|----------------------|-----------------------|-------------------------------------|
+| `blocks_by_hash`     | `BlockHeaderHash`     | `Block`                             |
+| `hash_by_height`     | `BE32(height)`        | `BlockHeaderHash`                   |
+| `tx_by_hash`         | `TransactionHash`     | `BlockHeaderHash || BE32(tx_index)` |
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
+
```

Comment:
How far back does the difficulty requirement calculation look? Could we do it based on the in-memory data (and if we did, how would we handle coming online after a restart, where there's no in-memory state)?

If we need to store work per tip I don't think it's a big deal, it's a small amount of data to store.

Separately from this issue, I'm wondering whether there's some plumbing we could do to get an estimate of the expected difficulty early in the verification process, with the motivation of trying reduce DoS risks by rejecting invalid blocks with a valid-but-too-low difficulty PoW as early as possible. I'm not sure whether this is possible.

Reply:
We just need to store the total work any time we could have multiple tips - because it is only used to choose between tips. But so we compare with `zcashd`, we should also store the total proof of work of the on-disk tip.

So here's what we'd store:
* total proof of work for the finalized tip (`u128` * 1)
* total proof of work for every un-finalized block (`u128` * ~100)

We need to store the total for every un-finalized block, because new blocks can fork off the chain at any point.

Reply:
It might be useful to add a sled tree `work_by_height` that stores the total proof of work for each height, so that the data we store is consistent between the final and non-final parts of the chain.